src/ResourceManagement/ResourceManager/Authentication/MSITokenProvider.cs

```diff
@@ -17,27 +18,44 @@ namespace Microsoft.Azure.Management.ResourceManager.Fluent.Authentication
     /// </summary>
     public class MSITokenProvider : ITokenProvider, IBeta
     {
-        private string resource;
-        private int? port;
+        private readonly string resource;
+        private readonly MSILoginInformation msiLoginInformation;
 
         /// <summary>
         /// Creates MSITokenProvider.
         /// </summary>
         /// <param name="resource">the resource to access using the MSI token</param>
-        /// <param name="port">the local endpoint port to retrieve the token from</param>
-        public MSITokenProvider(string resource, int? port = 50342)
+        /// <param name="msiLoginInformation">describes the managed service identity configuration</param>
+        public MSITokenProvider(string resource, MSILoginInformation msiLoginInformation)
         {
-            this.resource = resource;
-            this.port = port;
+            this.resource = resource ?? throw new ArgumentNullException("resource");
+            this.msiLoginInformation = msiLoginInformation ?? throw new ArgumentNullException("msiLoginInformation");
+            this.msiLoginInformation = msiLoginInformation;
         }
 
         public async Task<AuthenticationHeaderValue> GetAuthenticationHeaderAsync(CancellationToken cancellationToken)
         {
+            int port = msiLoginInformation.Port == null ? 50342 : msiLoginInformation.Port.Value;
             HttpRequestMessage msiRequest = new HttpRequestMessage(HttpMethod.Post, $"http://localhost:{port}/oauth2/token");
             msiRequest.Headers.Add("Metadata", "true");
 
-            IDictionary<string, string> parameters = new Dictionary<string, string>();
-            parameters.Add("resource", $"{resource}");
+            IDictionary<string, string> parameters = new Dictionary<string, string>
+            {
+                { "resource", $"{resource}" }
+            };
+            if (this.msiLoginInformation.UserAssignedIdentityObjectId != null)
+            {
+                parameters.Add("object_id", this.msiLoginInformation.UserAssignedIdentityObjectId);
+            }
+            else if (this.msiLoginInformation.UserAssignedIdentityClientId != null)
+            {
+                parameters.Add("client_id", this.msiLoginInformation.UserAssignedIdentityClientId);
+            }
+            else if (this.msiLoginInformation.UserAssignedIdentityResourceId != null)
+            {
+                parameters.Add("msi_res_id", this.msiLoginInformation.UserAssignedIdentityResourceId);
+            }
+
```

Comment:
Is it correct that only one parameter can be specified?

Reply:
@lenala these are mutually exclusive so we need if..elseif..elseif..